src/main/java/com/microsoft/azure/vmagent/AzureVMCloud.java

```diff
@@ -1004,6 +1013,26 @@ private static boolean isNodeEligibleForReuse(AzureVMAgent agentNode, AzureVMAge
         return false;
     }
 
+    private Object getLockForAgent(AzureVMAgent agent) {
+        synchronized (agentLocks) {
+            AtomicInteger lockCount = agentLocks.computeIfAbsent(agent, (a) -> new AtomicInteger(0));
+            lockCount.incrementAndGet();
```

Comment:
Why do we need an AtomicInteger instead of an object here?

Reply:
I use the AtomicInteger because it is a mutable Integer. 

When multiple threads want to acquire the lock I can give them the same object (=mutable Integer) while counting how many threads holding the lock for the agent. So when they release the lock the map can be cleared when all threads are done.

You could, of course, return an object but then another way of counting the threads would need to be implemented. (Like a custom map value which holds the lock object and the number of threads)